lib/api/completeMultipartUpload.js

```diff
@@ -363,66 +365,88 @@ function completeMultipartUpload(authInfo, request, log, callback) {
-                mpuBucket, mpuOverviewKey, aggregateETag,
-                storedPartsAsObjects, objMD, extraPartLocations);
-        },
-        function waterfall7(destinationBucket, dataLocations,
-                            metaStoreParams, mpuBucket, mpuOverviewKey,
-                            aggregateETag, storedPartsAsObjects, objMD,
-                            extraPartLocations, next) {
             const serverSideEncryption =
-                      destinationBucket.getServerSideEncryption();
+                      destBucket.getServerSideEncryption();
             let pseudoCipherBundle = null;
             if (serverSideEncryption) {
                 pseudoCipherBundle = {
-                    algorithm: destinationBucket.getSseAlgorithm(),
-                    masterKeyId: destinationBucket.getSseMasterKeyId(),
+                    algorithm: destBucket.getSseAlgorithm(),
+                    masterKeyId: destBucket.getSseMasterKeyId(),
                 };
             }
-            services.metadataStoreObject(destinationBucket.getName(),
-                dataLocations, pseudoCipherBundle, metaStoreParams, err => {
+            return services.versioningPreprocessing(bucketName,
+                destBucket, objectKey, objMD, null, log, (err, options) => {
+                    if (err) {
+                        return next(err, destBucket);
+                    }
+                    const deleteData = options.deleteData;
+                    metaStoreParams.versionId = options.versionId;
+                    metaStoreParams.versioning = options.versioning;
+                    metaStoreParams.isNull = options.isNull;
+                    metaStoreParams.nullVersionId = options.nullVersionId;
+                    return next(null, destBucket, dataLocations,
+                        metaStoreParams, mpuBucket, mpuOverviewKey,
+                        aggregateETag, storedPartsAsObjects, objMD,
+                        extraPartLocations, pseudoCipherBundle, deleteData);
+                });
+        },
+        function storeAsNewObj(destinationBucket, dataLocations,
+            metaStoreParams, mpuBucket, mpuOverviewKey, aggregateETag,
+            storedPartsAsObjects, objMD, extraPartLocations, pseudoCipherBundle,
+            deleteData, next) {
+            return services.metadataStoreObject(destinationBucket.getName(),
+                dataLocations, pseudoCipherBundle, metaStoreParams,
+                (err, res) => {
                     if (err) {
                         return next(err, destinationBucket);
                     }
-                    if (objMD && objMD.location) {
+                    const generatedVersionId = res ? res.versionId : undefined;
+                    // in cases where completing mpu overwrites a previous
+                    // null version or versioning is not enabled, need to delete
+                    // pre-existing data
```

Comment:
I think this should be in cases where versioning has never been unable or is suspended